### PR TITLE
Provide Static Metric Macros

### PR DESCRIPTION
**Pull request by breeswish**
Based on previous design proposals.

The code is based on PR #158 so that it will be cleaner after #158 is merged.

### CODE REVIEW

**Comment by breeswish** [comment]
@overvenus @siddontang PTAL

**Comment by siddontang** [comment]
why using another crate for it?

**Comment by breeswish** [comment]
It is a limitation by rust compiler for now that crates marked as containing procedural macros cannot export other kind of things, like normal macros, functions, etc, so that this procedural macro stays in another crate, created specially for it.

**Comment by breeswish** [comment]
Updated base branch to v0.4, and the diff is much cleaner now. PTAL  @siddontang @overvenus

**Comment by breeswish** [comment]
any updates? @overvenus

File: static-metric/LICENSE [hunk 1]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (siddontang) [inline static-metric/LICENSE]: No need including LICENSE here.

**Comment by breeswish** [comment]
@siddontang License removed

**Comment by siddontang** [comment]
Please give a benchmark.

**Comment by siddontang** [comment]
PTAL @Hoverbear

**Comment by breeswish** [comment]
@siddontang It is compile-time macros, does not introduce runtime overheads. It should be as fast as hand write static metrics.

**Comment by siddontang** [comment]
@breeswish 

No matter static or not, we need to provide a benchmark definitely.

**Comment by breeswish** [comment]
wired.. I didn't close this PR..

**Comment by breeswish** [comment]
rebase to master to open this PR again

**Pull request closed without merging**